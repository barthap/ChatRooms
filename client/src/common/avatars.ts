export function userAvatarUrl(userHash: string) {
  return `https://identicon-api.herokuapp.com/${encodeURIComponent(userHash)}/50?format=png`;
}

function stringToHslColor(str: string, s: number = 30, l: number = 80): string {
  let hash = 0;
  for (let i = 0; i < str.length; i++) {
    hash = str.charCodeAt(i) + ((hash << 5) - hash);
  }

  hash = hash >= 0 ? hash : -hash;
  const h = hash % 360;
  //return 'hsl(' + h + ', ' + s + '%, ' + l + '%)';
  return hslToHex(h, s, l);
}

function hslToHex(h: number, s: number, l: number): string {
  l /= 100;
  const a = (s * Math.min(l, 1 - l)) / 100;
  const f = (n: number) => {
    const k = (n + h / 30) % 12;
    const color = l - a * Math.max(Math.min(k - 3, 9 - k, 1), -1);
    return Math.round(255 * color)
      .toString(16)
      .padStart(2, '0'); // convert to Hex and prefix "0" if needed
  };
  return `#${f(0)}${f(8)}${f(4)}`;
}

export function groupAvatarUrl(groupName: string) {
  const color = stringToHslColor(groupName).substring(1);
  return `https://eu.ui-avatars.com/api/?background=${color}&size=50&name=${encodeURIComponent(
    groupName
  )}`;
}

export function groupAvatarUrl2(groupName: string) {
  return `https://avatar.oxro.io/avatar.png?name=${encodeURIComponent(groupName)}`;
}

export function personAvatarUrl(nr: number) {
  return `https://i.pravatar.cc/100?a=${nr}`;
}
