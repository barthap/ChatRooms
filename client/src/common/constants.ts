export const WEBSOCKET_URL = process.env.REACT_APP_WS_URL || '';
export const API_URL = process.env.REACT_APP_API_URL || '/api';
