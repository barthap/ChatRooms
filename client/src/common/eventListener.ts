type EventListener<TEventArgs extends unknown[]> = (...args: TEventArgs) => void;
type ListenerId = number & { readonly _: unique symbol };

export class EventHandler<TEventArgs extends unknown[] = []> {
  private listeners: Map<ListenerId, EventListener<TEventArgs>> = new Map();

  addListener(listener: EventListener<TEventArgs>) {
    const id = this.listeners.size as ListenerId;
    this.listeners.set(id, listener);
  }

  removeListener(id: ListenerId) {
    this.listeners.delete(id);
  }

  notify(...args: TEventArgs) {
    for (const l of this.listeners.values()) {
      l(...args);
    }
  }

  removeAllListeners() {
    this.listeners.clear();
  }
}
