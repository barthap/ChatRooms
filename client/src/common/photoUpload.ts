const IMGBB_API_ENDPOINT_URL = 'https://api.imgbb.com/1/upload';
const IMGBB_API_KEY = process.env.REACT_APP_IMGBB_KEY;

const DEFAULT_EXPIRATION_TIME = 60 * 60 * 24; // 24 hours
const IMAGE_EXPIRATION_TIME =
  process.env.REACT_APP_IMAGE_EXPIRE_TIME || DEFAULT_EXPIRATION_TIME.toString();

export interface PhotoUploadResponse {
  data: {
    id: string;
    title: string;
    url_viewer: string;
    url: string;
    display_url: string;
    size: number;
    time: string | number;
    expiration: string | number;
    image?: {
      filename: string;
      name: string;
      mime: string;
      extension: string;
      url: string;
    };
    thumb?: {
      filename: string;
      name: string;
      mime: string;
      extension: string;
      url: string;
    };
    delete_url: string;
  };
  success: boolean;
  status: number;
}

export async function uploadImage(image: File): Promise<PhotoUploadResponse> {
  if (!IMGBB_API_KEY) {
    throw new Error('Image hosting API key is not defined! Please define REACT_APP_IMGBB_KEY');
  }

  const formData = new FormData();
  formData.append('image', image);
  formData.append('key', IMGBB_API_KEY);
  formData.append('expiration', IMAGE_EXPIRATION_TIME);

  try {
    const response = await fetch(IMGBB_API_ENDPOINT_URL, {
      method: 'POST',
      body: formData,
    });

    const body = (await response.json()) as PhotoUploadResponse;
    if (body.success) {
      return body;
    }
    throw new Error('Uploading image failed: ' + JSON.stringify(body));
  } catch (e) {
    throw new Error('Error when uploading image: ' + e);
  }
}
