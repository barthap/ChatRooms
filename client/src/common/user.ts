export interface IUser {
  name: string;
  id: string;
  session_id: string;
}
