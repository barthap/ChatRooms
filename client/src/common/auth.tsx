import React, { useContext, createContext } from 'react';
import { useCookies } from 'react-cookie';

import { API_URL } from './constants';
import { IUser } from './user';

async function createUser(username: string): Promise<IUser> {
  const response = await fetch(`${API_URL}/users/`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      username,
    }),
  });

  const body = await response.json();
  if (response.status === 201) {
    return body;
  }
  throw new Error(body.message ?? 'unknown error: ' + JSON.stringify(body));
}

async function deleteUser(userId: string): Promise<void> {
  const response = await fetch(`${API_URL}/users/${userId}`, {
    method: 'DELETE',
  });

  if (response.status !== 204) {
    throw new Error('Error when deleting user: ' + (await response.text()));
  }
}

interface IAuth {
  user: IUser | null;
  isAuthenticated: boolean;
  signIn: (username: string) => Promise<IUser>;
  signOut: () => Promise<void>;
  invalidateSession: () => void;
}

/** For more details on
 * `authContext`, `ProvideAuth`, `useAuth` and `useProvideAuth`
 * refer to: https://usehooks.com/useAuth/
 */
const authContext = createContext<IAuth | undefined>(undefined);

export function ProvideAuth({ children }: React.PropsWithChildren<object>) {
  const auth = useProvideAuth();
  return <authContext.Provider value={auth}>{children}</authContext.Provider>;
}

export function useAuth() {
  return useContext(authContext);
}

function useProvideAuth(): IAuth {
  const [cookies, setCookie, removeCookie] = useCookies(['user']);
  const user: IUser | null = cookies.user ?? null;

  console.log('hook', user);

  const signIn = async (username: string) => {
    const user = await createUser(username);
    setCookie('user', user, { maxAge: 24 * 60 * 60 });
    return user;
  };

  const signOut = async () => {
    invalidateSession();
    user && (await deleteUser(user.id));
  };

  const invalidateSession = () => {
    removeCookie('user');
  };

  return {
    isAuthenticated: user != null,
    user,
    signIn,
    signOut,
    invalidateSession,
  };
}
