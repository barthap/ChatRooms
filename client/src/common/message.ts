import { IUser } from './user';

export enum MessageType {
  MESSAGE,
  USER_JOINED,
  USER_LEFT,
}

interface IMessageBase {
  id: string;
  type: MessageType;
  timestamp: number;
}

interface MessageWithContent extends IMessageBase {
  type: typeof MessageType.MESSAGE;
  sender: IUser;
}

export interface IFileMessage extends MessageWithContent {
  url: string;
}
export interface ITextMessage extends MessageWithContent {
  content: string;
}

interface IUserJoinedMessage extends IMessageBase {
  type: typeof MessageType.USER_JOINED;
  user: IUser;
}

interface IUserLeftMessage extends IMessageBase {
  type: typeof MessageType.USER_LEFT;
  user: IUser;
}

export function isTextMessage(msg?: IMessage): msg is ITextMessage {
  return msg?.type === MessageType.MESSAGE && (msg as any).content;
}

export function isFileMessage(msg?: IMessage): msg is IFileMessage {
  return msg?.type === MessageType.MESSAGE && (msg as any).url;
}

export type IMessage = ITextMessage | IFileMessage | IUserJoinedMessage | IUserLeftMessage;
