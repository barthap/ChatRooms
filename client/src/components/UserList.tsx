import { Status, StatusList } from '@chatscope/chat-ui-kit-react';
import React from 'react';
import '@chatscope/chat-ui-kit-styles/dist/default/styles.min.css';
import 'bootstrap/dist/css/bootstrap.min.css';
import { OverlayTrigger, Tooltip } from 'react-bootstrap';

import { IUser } from '../common/user';

type UserStatus = 'available' | 'eager' | 'dnd' | 'unavailable' | 'invisible' | 'away';

function getStatusText(status: UserStatus, roomName?: string) {
  switch (status) {
    case 'available':
      return 'In your room';
    case 'dnd':
      return 'Offline';
    case 'invisible':
      return 'Outside any room';
    case 'away':
      return `In room: ${roomName}`;
    default:
      // this should never happen
      return 'Unknown status';
  }
}

function renderUser(user: IUser, currentUser?: IUser) {
  const currentRoom = currentUser?.current_room;
  const isOffline = user.session_id == null;
  const isCurrentRoom = user.current_room?.id === currentRoom?.id;
  const isSelected = user.id === currentUser?.id;

  const status: UserStatus = isOffline
    ? 'dnd'
    : currentRoom == null
    ? 'invisible'
    : isCurrentRoom
    ? 'available'
    : 'away';

  const statusText = isSelected ? 'You' : getStatusText(status, user.current_room?.name);

  return (
    <OverlayTrigger
      // @ts-expect-error
      as={Status}
      key={user.id}
      placement="left"
      overlay={<Tooltip id={`tooltip-${user.id}`}>{statusText}</Tooltip>}>
      <Status key={user.id} status={status} name={user.name} selected={isSelected} />
    </OverlayTrigger>
  );
}

export default function UserList({ users, currentUser }: { users: IUser[]; currentUser?: IUser }) {
  const renderUserForCurrent = (user: IUser) => renderUser(user, currentUser);
  return (
    <StatusList size="md" style={{ padding: 0 }}>
      {users.map(renderUserForCurrent)}
    </StatusList>
  );
}
