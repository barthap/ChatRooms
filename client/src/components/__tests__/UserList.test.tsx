import { render, screen } from '@testing-library/react';
import React from 'react';

import { createTestRoom, createTestUser } from '../../__tests__/testUtils';
import { IUser } from '../../common/user';
import UserList from '../UserList';

const testRoom1 = createTestRoom('Room 1');
const testRoom2 = createTestRoom('Room 2');

const testUserList: IUser[] = [
  createTestUser('Janusz', { online: true, room: testRoom1 }),
  createTestUser('Grazyna', { online: true, room: testRoom1 }),
  createTestUser('Pioter', { online: true, room: testRoom2 }),
  createTestUser('Brajan', { online: false }),
];
const janusz = testUserList[0];

describe('<UserList />', () => {
  it('displays user status list', () => {
    const userList = render(<UserList users={testUserList} currentUser={janusz} />);
    expect(userList).toMatchSnapshot();
  });
  it('displays correct status', () => {
    render(<UserList users={testUserList} currentUser={janusz} />);

    expect(getStatusClassName('Grazyna')).toContain('available');
    expect(getStatusClassName('Pioter')).toContain('away');
    expect(getStatusClassName('Brajan')).toContain('dnd');
  });
});

const getStatusClassName = (username: string) =>
  screen.getByText(username).parentElement?.className;
