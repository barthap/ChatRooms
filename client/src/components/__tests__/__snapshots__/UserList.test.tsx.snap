// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserList /> displays user status list 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <ul
        class="cs-status-list cs-status-list--md"
        style="padding: 0px;"
      >
        <li>
          <div
            aria-selected=""
            class="cs-status cs-status--md cs-status--available cs-status--selected cs-status--named"
          >
            <div
              class="cs-status__bullet"
            />
            <div
              class="cs-status__name"
            >
              Janusz
            </div>
          </div>
        </li>
        <li>
          <div
            class="cs-status cs-status--md cs-status--available cs-status--named"
          >
            <div
              class="cs-status__bullet"
            />
            <div
              class="cs-status__name"
            >
              Grazyna
            </div>
          </div>
        </li>
        <li>
          <div
            class="cs-status cs-status--md cs-status--away cs-status--named"
          >
            <div
              class="cs-status__bullet"
            />
            <div
              class="cs-status__name"
            >
              Pioter
            </div>
          </div>
        </li>
        <li>
          <div
            class="cs-status cs-status--md cs-status--dnd cs-status--named"
          >
            <div
              class="cs-status__bullet"
            />
            <div
              class="cs-status__name"
            >
              Brajan
            </div>
          </div>
        </li>
      </ul>
    </div>
  </body>,
  "container": <div>
    <ul
      class="cs-status-list cs-status-list--md"
      style="padding: 0px;"
    >
      <li>
        <div
          aria-selected=""
          class="cs-status cs-status--md cs-status--available cs-status--selected cs-status--named"
        >
          <div
            class="cs-status__bullet"
          />
          <div
            class="cs-status__name"
          >
            Janusz
          </div>
        </div>
      </li>
      <li>
        <div
          class="cs-status cs-status--md cs-status--available cs-status--named"
        >
          <div
            class="cs-status__bullet"
          />
          <div
            class="cs-status__name"
          >
            Grazyna
          </div>
        </div>
      </li>
      <li>
        <div
          class="cs-status cs-status--md cs-status--away cs-status--named"
        >
          <div
            class="cs-status__bullet"
          />
          <div
            class="cs-status__name"
          >
            Pioter
          </div>
        </div>
      </li>
      <li>
        <div
          class="cs-status cs-status--md cs-status--dnd cs-status--named"
        >
          <div
            class="cs-status__bullet"
          />
          <div
            class="cs-status__name"
          >
            Brajan
          </div>
        </div>
      </li>
    </ul>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
